### PR TITLE
[ui][new] kudu connector test

### PR DESCRIPTION
**Pull request by mcada**
//test: `@integrations-kudu`

### CODE REVIEW

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by avano** [comment]
retest this please

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by mcada** [comment]
Looks like 10 minutes timeout was not enough for kudu server to deploy, increasing timeout to 15.

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by mcada** [comment]
retest this please

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by avano** [comment]
retest this please

File: ui-tests/src/test/java/io/syndesis/qe/steps/other/KuduSteps.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            fail("Table creation failed: " + response);
+        }
+    }
+
+    @When("^delete table from Kudu server$")
+    public void deleteTable() {
+        String response = HttpUtils.doGetRequest("http://" + OpenShiftUtils.getInstance().getRoutes().stream()
+                .filter(route -> "kudu-rest-api".equals(route.getMetadata().getName()))
+                .findFirst().get().getSpec().getHost() + "/kudu/table/delete").getBody();
+        log.info("Table deletion log: " + response);
```
Comment (avano) [inline ui-tests/src/test/java/io/syndesis/qe/steps/other/KuduSteps.java]: needed? if yes probably lower to debug, wdyt?

File: ui-tests/src/test/resources/features/integrations/kudu.feature [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      | Apache Kudu | kudu | KuduConnector | description |
+    And navigate to the "Home" page
+
+#
+#  1. Insert data into kudu table
+#
+  @ui
+  @kudu
+  @database
+  @datamapper
```
Comment (avano) [inline ui-tests/src/test/resources/features/integrations/kudu.feature]: are these tags needed when you have the whole feature annotated?

File: utilities/src/main/java/io/syndesis/qe/TestSuiteParent.java [hunk 3]
```diff
@@ -50,6 +54,13 @@ public static void lockNamespace() throws InterruptedException {
 
     @AfterClass
     public static void tearDown() {
+
+        if(KUDU_WAS_USED) {
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/TestSuiteParent.java]: isn't `TestUtils.isDcDeployed()` enough?

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java [hunk 4]
```diff
@@ -52,9 +53,27 @@ public void sendAndRecieveMessage(String senderTopic, String receiverTopic) {
         }
 
         //check flag
-        Assertions.assertThat(Receiver.RECEIVED_FLAG).as("Message was not received!!!")
+        assertThat(Receiver.RECEIVED_FLAG).as("Message was not received!!!")
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java]: one exclamation mark should be enough :-) (I know you didn't write that message)

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java [hunk 5]
```diff
@@ -52,9 +53,27 @@ public void sendAndRecieveMessage(String senderTopic, String receiverTopic) {
...
                 .isEqualTo(1);
+    }
+
+    @When("^send mqtt message to \"([^\"]*)\" topic$")
+    public void sendMqttMessage(String senderTopic) {
 
+        try {
+            portForward();
+            // give port forward some time
+            TestUtils.sleepIgnoreInterrupt(2000);
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java]: is this really needed?

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java [hunk 6]
```diff
@@ -52,9 +53,27 @@ public void sendAndRecieveMessage(String senderTopic, String receiverTopic) {
...
+
+            //send message via client
+            mqttUtils.sendMessage("{\"key\" : 1,\"value\" : \"FirstValue\"}", senderTopic);
+
+            //give it some short time for delivery
+            TestUtils.sleepIgnoreInterrupt(4000);
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java]: this should probably be in the consumer with some intelligent wait

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java [hunk 7]
```diff
@@ -36,6 +38,22 @@ public void deployFTPServer() {
         FtpTemplate.deploy();
     }
 
+    @Given("^deploy Kudu rest API")
+    public void deployKuduRestAPI() { KuduRestAPITemplate.deploy(); }
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: format

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java [hunk 8]
```diff
@@ -36,6 +38,22 @@ public void deployFTPServer() {
...
+
+    @Given("^deploy Kudu server")
+    public void deployKuduServer() { KuduTemplate.deploy(); }
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: format

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java (hunk 7 above)
Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: + end with $ in the annotation

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java (hunk 8 above)
Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: + end with $ in the annotation

File: utilities/src/main/java/io/syndesis/qe/templates/KuduRestAPITemplate.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    public static void cleanUp() {
+        OpenShiftUtils.getInstance().getDeploymentConfigs().stream().filter(dc -> dc.getMetadata().getName().equals(APP_NAME)).findFirst()
+                .ifPresent(dc -> OpenShiftUtils.getInstance().deleteDeploymentConfig(dc, true));
+        OpenShiftUtils.getInstance().getServices().stream().filter(service -> APP_NAME.equals(service.getMetadata().getName())).findFirst()
+                .ifPresent(service -> OpenShiftUtils.getInstance().deleteService(service));
+        OpenShiftUtils.getInstance().getRoutes().stream().filter(route -> APP_NAME.equals(route.getMetadata().getName())).findFirst()
+                .ifPresent(route -> OpenShiftUtils.getInstance().deleteRoute(route));
+        TestUtils.sleepIgnoreInterrupt(5 * 1000L);
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/templates/KuduRestAPITemplate.java]: needed?

File: utilities/src/main/java/io/syndesis/qe/templates/KuduTemplate.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+@Slf4j
+public class KuduTemplate {
+    private static final String APP_NAME = "syndesis-kudu";
+    private static final String ROUTE_NAME = "kudu";
+    private static final String LABEL_NAME = "syndesis.io/component";
+    //Template without persistent volumes, works on both minishift and openshift
+    private static final String TEMPLATE_URL = "https://gist.githubusercontent.com/mcada/4778b292836a4aaca1f274d0b13cae38/raw/b079c9c5950b5b0a6dd87d253bbb47d2fc734d94/kudu-final";
+    //Template with persistent volumes, works only on minishift
+    //private static final String TEMPLATE_URL = "https://raw.githubusercontent.com/syndesisio/syndesis/92bd31e41ff66edbd0e6b47738f3da54930bd295/install/extras/syndesis-kudu.yml";
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/templates/KuduTemplate.java]: needed?

File: utilities/src/main/java/io/syndesis/qe/templates/KuduTemplate.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import io.syndesis.qe.wait.OpenShiftWaitUtils;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class KuduTemplate {
+    private static final String APP_NAME = "syndesis-kudu";
+    private static final String ROUTE_NAME = "kudu";
+    private static final String LABEL_NAME = "syndesis.io/component";
+    //Template without persistent volumes, works on both minishift and openshift
+    private static final String TEMPLATE_URL = "https://gist.githubusercontent.com/mcada/4778b292836a4aaca1f274d0b13cae38/raw/b079c9c5950b5b0a6dd87d253bbb47d2fc734d94/kudu-final";
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/templates/KuduTemplate.java]: add the template directly into the repository

File: utilities/src/main/java/io/syndesis/qe/templates/KuduTemplate.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        OpenShiftUtils.getInstance().getDeploymentConfigs().stream().filter(dc -> dc.getMetadata().getName().equals(APP_NAME)).findFirst()
+                .ifPresent(dc -> OpenShiftUtils.getInstance().deleteDeploymentConfig(dc, true));
+        OpenShiftUtils.getInstance().getServices().stream().filter(service -> APP_NAME.equals(service.getMetadata().getName())).findFirst()
+                .ifPresent(service -> OpenShiftUtils.getInstance().deleteService(service));
+        OpenShiftUtils.getInstance().getRoutes().stream().filter(route -> ROUTE_NAME.equals(route.getMetadata().getName())).findFirst()
+                .ifPresent(route -> OpenShiftUtils.getInstance().deleteRoute(route));
+        OpenShiftUtils.getInstance().getPersistentVolumeClaims().stream()
+                .filter(volume -> (volume.getMetadata().getName()).contains(APP_NAME))
+                .forEach(volume -> OpenShiftUtils.getInstance().deletePersistentVolumeClaim(volume));
+        TestUtils.sleepIgnoreInterrupt(5 * 1000L);
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/templates/KuduTemplate.java]: needed?

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java (hunk 4 above)
Comment (mcada) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java]: Actually I did... :D

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java (hunk 5 above)
Comment (mcada) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/MqttValidationSteps.java]: Yes, moved to appropriate place. It failed a few times, better +2sec than unstable tests.

File: utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java (hunk 8 above)
Comment (mcada) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: Hmm, the format is the same as other methods in the class, running it through auto format makes some inconsistencies but at least they will be consistent through all the tests ^^

Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: idea shows it in single line even if the source file is properly formatted in multiple lines

Comment (mcada) [reply utilities/src/main/java/io/syndesis/qe/bdd/validation/OpenshiftValidationSteps.java]: Uhmm... True :) But some stay in long format, no idea why... Anyway it is redone as you wanted it to be :)

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by avano** [comment]
retest this please

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

File: utilities/src/main/resources/templates/syndesis-kudu.yml [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    triggers:
+    - imageChangeParams:
+        automatic: true
+        containerNames:
+        - syndesis-kudu
+        from:
+          kind: ImageStreamTag
+          name: syndesis-kudu:latest
+      type: ImageChange
+    - type: ConfigChange
```
Comment (avano) [inline utilities/src/main/resources/templates/syndesis-kudu.yml]: add a new line here

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Comment by syndesisqe** [comment]
Build finished. Tests FAILED

**Pull request merged**